File: 2023/06/c/main.c
#include "main.h"
#include "run.h"
#include "utils_num.h"
#include "utils_str.h"
#include <math.h>
#include <stdlib.h>
#include <string.h>

static const int32_t HAS_ALTERNATE = 0;

static int64_t ohnomath(const int64_t b, const int64_t c) {
   double min = floor((-b - sqrt(b * b - 4 * c)) / -2 - 0.001);
   double max = ceil((-b + sqrt(b * b - 4 * c)) / -2 + 0.001);
   return min - max + 1;
}

static int64_t *getnum(const char *restrict str, size_t *count) {
   size_t chonkSz, i, j;
   int64_t *ary;
   char **chonk;

   *count = 0;
   chonk = str_splitc(str, ' ', &chonkSz);
   for (i = 0; i < chonkSz; i++)
      if (strlen(chonk[i]))
         (*count)++;

   ary = malloc(*count * sizeof(*ary));
   for (i = 0, j = 0; i < chonkSz; i++) {
      if (strlen(chonk[i]))
         ary[j++] = atoll(chonk[i]);
      free(chonk[i]);
   }
   free(chonk);

   return ary;
}

static int64_t getnumjoin(const char *restrict str) {
   size_t chonkSz, i;
   int64_t res;
   char **chonk;

   chonk = str_splitc(str, ' ', &chonkSz);
   res = 0;
   for (i = 0; i < chonkSz; i++) {
      if (i)
         res *= pow(10, (int64_t)strlen(chonk[i]));
      res += atoll(chonk[i]);
      free(chonk[i]);
   }
   free(chonk);

   return res;
}

static char *part1(const char *restrict input, const int32_t isTest) {
   size_t chunksSz, linesSz, timesSz, distancesSz, i;
   int64_t *times, *distances, res;
   char **chunks, **lines;

   lines = str_splitc(input, '\n', &linesSz);
   chunks = str_splitc(lines[0], ':', &chunksSz);
   times = getnum(chunks[1], &timesSz);
   free(chunks[0]);
   free(chunks[1]);
   free(chunks);

   chunks = str_splitc(lines[1], ':', &chunksSz);
   distances = getnum(chunks[1], &distancesSz);
   free(chunks[0]);
   free(chunks[1]);
   free(chunks);

   free(lines[0]);
   free(lines[1]);
   free(lines);

   res = 1;
   for (i = 0; i < timesSz; i++)
      res *= ohnomath(times[i], distances[i]);

   free(times);
   free(distances);

   return num_tostr(res);
}

static char *part2(const char *restrict input, const int32_t isTest) {
   size_t chunksSz, linesSz;
   int64_t time, distance;
   char **chunks, **lines;

   lines = str_splitc(input, '\n', &linesSz);
   chunks = str_splitc(lines[0], ':', &chunksSz);
   time = getnumjoin(chunks[1]);
   free(chunks[0]);
   free(chunks[1]);
   free(chunks);

   chunks = str_splitc(lines[1], ':', &chunksSz);
   distance = getnumjoin(chunks[1]);
   free(chunks[0]);
   free(chunks[1]);
   free(chunks);

   free(lines[0]);
   free(lines[1]);
   free(lines);

   return num_tostr(ohnomath(time, distance));
}

int main(int argc, char *argv[]) {
   return run(argc, argv, part1, part2, HAS_ALTERNATE);
}

File: 2023/06/java/Main.java
package kival.aoc;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

import kival.aoc.core.*;

public class Main {
   public final boolean HAS_IO = false;
   public final boolean HAS_ALTERNATE = false;

   private long ohnomath(long b, long c) {
      double pepsilon = 0.001;
      double mn = Math.floor((b + Math.sqrt(Math.abs(b * b - 4 * c))) / 2 - pepsilon);
      double mx = Math.ceil((b - Math.sqrt(Math.abs(b * b - 4 * c))) / 2 + pepsilon);
      return (long) (mn - mx + 1);
   }
   
   public String part1(String input,boolean isTest) {
      long res = 1;
      long[][] td = input.lines()
         .map(x -> x.split(":")[1].trim().split(" "))
         .map(x -> Stream.of(x).filter(y -> !y.isEmpty()).mapToLong(Long::parseLong).toArray())
         .toArray(long[][]::new);
      for (int i = 0; i < td[0].length; i++) res *= ohnomath(td[0][i], td[1][i]);
      return String.valueOf(res);
   }

   public String part2(String input,boolean isTest) {
      long[] td = input.lines()
         .map(x -> String.join("", x.split(":")[1].trim().split(" ")))
         .mapToLong(Long::parseLong)
         .toArray();
      return String.valueOf(ohnomath(td[0], td[1]));
   }

   public void main(String[] args) {
      Run.execute(args, this::part1, this::part2, HAS_ALTERNATE, HAS_IO);
   }
}